# 3. **F1 Win Prediction Project - Data Preprocessing**

# Table of Contents
1. [Introduction](#introduction)
2. [Data Types and Integrity Checks](#data-types-and-integrity-checks)
    - [Data Types](#data-types)
    - [Missing Data](#missing-data)
    - [Duplicate Rows](#duplicate-rows)
    - [Uniqueness of Data](#uniqueness-of-data)
    - [Data Range](#data-range)
    - [Consistency](#consistency)
    - [Correctness](#correctness)
    - [Data Integrity](#data-integrity)
3. [Data Preprocessing](#data-preprocessing)
    - [Importing Libraries and Notebook Setup](#importing-libraries-and-notebook-setup)
    - [Initial DataFrame Information](#initial-dataframe-information)
    - [Creating New Index Column](#creating-new-index-column)
    - [Rename Columns](#rename-columns)
    - [Drop Redundant Columns](#drop-redundant-columns)
    - [Changing Data Types](#changing-data-types)
    - [Dropping Duplicates](#dropping-duplicates)
    - [Handling Unreasonable Data Ranges](#handling-unreasonable-data-ranges)
4. [Feature Engineering / Transformation](#feature-engineering--transformation)
    - [Constructor Points at Stage of Season](#constructor-points-at-stage-of-season)
    - [Driver Points at Stage of Season](#driver-points-at-stage-of-season)
    - [Other Feature Engineering Ideas](#other-feature-engineering-ideas)
5. [One Hot Encoding](#one-hot-encoding)
6. [Reviewing 'fastest_lap_from_last_race' Column](#reviewing-fastest_lap_from_last_race-column)
7. [Exporting the Processed DataFrame](#exporting-the-processed-dataframe)


### **Areas to Fix**

- **Data Types**: Ensure that each column has the appropriate data type for the kind of data it contains. For instance, categorical data should not be typed as numeric and vice versa. If any columns are meant to be categorical or date/time but are currently recognized as 'object' or 'int64', they should be converted to the proper data type.

- **Missing Data**: The dataset contains columns with high percentages of missing values, such as 'laps_in_previous_race' and 'laps_completed_in_previous_races'. I will need to decide how to handle these, whether by imputation, deletion, or acquisition of more data if possible. For columns with a small amount of missing data, imputation might be feasible, while for those with a large percentage, it might be more appropriate to consider dropping the column.

- **Duplicate Rows**: I will also check for any duplicate rows that might skew the analysis. If duplicates are not meaningful for the study, they will be removed.

- **Data Range**: Next, I will verify the range of values in numerical columns. For instance, if 'year' has a minimum value that's in the future or a past date that's not plausible (e.g. outside the date ranges I collected), these could be data entry/collection errors. 

- **Consistency**: Ensure that the data is consistent throughout the dataset. For example, if 'country' and 'nationality_of_circuit' are very similar and basically  represent the same information, so they should possibly merged (or dropped).

- **Correctness**: For columns with 'inf' values for uniqueness, ensure they are correctly calculated. An 'inf' value might indicate a division by zero error, suggesting that the column might be entirely unique or entirely composed of a single value, each of which has different implications.

- **Normalization/Standardization**: For machine learning purposes, I may need to standardize or normalize numerical data to ensure that the scale of the data does not unduly influence the model. This may be done at the modelling stage, rather than here, but it's a consideration I need to think about.

### Importing Libraries and Notebook Setup

In [22]:
# Install libraries
import pandas as pd
import numpy as np
from datetime import datetime
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats

In [23]:
process_df = pd.read_csv('C:/Users/Alex/OneDrive/BrainStation/Data_Science_Bootcamp/Capstone_Project/capstone-Aboard89/data/data_analysis.csv')

In [24]:
process_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11235 entries, 0 to 11234
Data columns (total 36 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Index                             11235 non-null  int64  
 1   resultId                          11235 non-null  int64  
 2   raceId                            11235 non-null  int64  
 3   year                              11235 non-null  int64  
 4   race                              11235 non-null  object 
 5   country                           11235 non-null  object 
 6   nationality_of_circuit            11235 non-null  object 
 7   driverId                          11235 non-null  int64  
 8   number                            11235 non-null  int64  
 9   driver_name                       11235 non-null  object 
 10  F2_champion                       11235 non-null  int64  
 11  Former_F1_World_Champion          11235 non-null  int64  
 12  Nati

In [25]:
pd.set_option('display.max_columns', None)
process_df.head()

Index  resultId  raceId  year                  race country  \
0      1      4721     240  1995  Brazilian Grand Prix  brazil   
1      1      4724     240  1995  Brazilian Grand Prix  brazil   
2      1      4746     240  1995  Brazilian Grand Prix  brazil   
3      1      4745     240  1995  Brazilian Grand Prix  brazil   
4      1      4744     240  1995  Brazilian Grand Prix  brazil   

  nationality_of_circuit  driverId  number            driver_name  \
0              Brazilian        30       1     Michael Schumacher   
1              Brazilian        57       8          Mika Häkkinen   
2              Brazilian        94      23      Pierluigi Martini   
3              Brazilian        44      26          Olivier Panis   
4              Brazilian        49      30  Heinz-Harald Frentzen   

   F2_champion  Former_F1_World_Champion Nationality  home_race  \
0            0                         1      German          0   
1            0                         0     Finnish          0   
2            0                         0     Italian          0   
3            0                         0      French          0   
4            0                         0      German          0   

   constructorId constructor engine_manufacturer constructor_nationality  \
0             22    Benetton             Renault                 Italian   
1              1     McLaren            Mercedes                 British   
2             18     Minardi                Ford                 Italian   
3             27      Ligier         Mugen-Honda                  French   
4             15      Sauber                Ford                   Swiss   

   number.1  starting_grid_position  positionOrder  points  \
0         1                       2              1    10.0   
1         8                       7              4     3.0   
2        23                      17             26     0.0   
3        26                      10             25     0.0   
4        30                      14             24     0.0   

   points_in_previous_race  laps  laps_completed_in_previous_races  \
0                      NaN    71                               NaN   
1                      NaN    70                               NaN   
2                      NaN     0                               NaN   
3                      NaN     0                               NaN   
4                      NaN    10                               NaN   

              time milliseconds fastestLap_ms fastest_lap_from_last_race  \
0  01:38:34.154000      5914154         81009                        NaN   
1               \N           \N     Not Found                        NaN   
2               \N           \N     Not Found                        NaN   
3               \N           \N     Not Found                        NaN   
4               \N           \N         84001                        NaN   

   number_of_pit_stops  average_time_lost_in_pits  statusId      status  \
0                    3                      31.83         1    Finished   
1                    0                       0.00        11      +1 Lap   
2                    0                       0.00         6     Gearbox   
3                    0                       0.00         4   Collision   
4                    0                       0.00        10  Electrical   

   major_regulation_change  laps_in_previous_race  race_win  
0                        0                    NaN         1  
1                        0                    NaN         0  
2                        0                    NaN         0  
3                        0                    NaN         0  
4                        0                    NaN         0

## Data Preprocessing

#### **New Index Column**

In [26]:
process_df['new_index'] = process_df['Index'] + process_df['driverId']

The above is create a new index column, so that I can track who the model think is going to win races when we get to the prediction stage. 

I then want to make this the new index of my dataframe

In [27]:
process_df.head()

Index  resultId  raceId  year                  race country  \
0      1      4721     240  1995  Brazilian Grand Prix  brazil   
1      1      4724     240  1995  Brazilian Grand Prix  brazil   
2      1      4746     240  1995  Brazilian Grand Prix  brazil   
3      1      4745     240  1995  Brazilian Grand Prix  brazil   
4      1      4744     240  1995  Brazilian Grand Prix  brazil   

  nationality_of_circuit  driverId  number            driver_name  \
0              Brazilian        30       1     Michael Schumacher   
1              Brazilian        57       8          Mika Häkkinen   
2              Brazilian        94      23      Pierluigi Martini   
3              Brazilian        44      26          Olivier Panis   
4              Brazilian        49      30  Heinz-Harald Frentzen   

   F2_champion  Former_F1_World_Champion Nationality  home_race  \
0            0                         1      German          0   
1            0                         0     Finnish          0   
2            0                         0     Italian          0   
3            0                         0      French          0   
4            0                         0      German          0   

   constructorId constructor engine_manufacturer constructor_nationality  \
0             22    Benetton             Renault                 Italian   
1              1     McLaren            Mercedes                 British   
2             18     Minardi                Ford                 Italian   
3             27      Ligier         Mugen-Honda                  French   
4             15      Sauber                Ford                   Swiss   

   number.1  starting_grid_position  positionOrder  points  \
0         1                       2              1    10.0   
1         8                       7              4     3.0   
2        23                      17             26     0.0   
3        26                      10             25     0.0   
4        30                      14             24     0.0   

   points_in_previous_race  laps  laps_completed_in_previous_races  \
0                      NaN    71                               NaN   
1                      NaN    70                               NaN   
2                      NaN     0                               NaN   
3                      NaN     0                               NaN   
4                      NaN    10                               NaN   

              time milliseconds fastestLap_ms fastest_lap_from_last_race  \
0  01:38:34.154000      5914154         81009                        NaN   
1               \N           \N     Not Found                        NaN   
2               \N           \N     Not Found                        NaN   
3               \N           \N     Not Found                        NaN   
4               \N           \N         84001                        NaN   

   number_of_pit_stops  average_time_lost_in_pits  statusId      status  \
0                    3                      31.83         1    Finished   
1                    0                       0.00        11      +1 Lap   
2                    0                       0.00         6     Gearbox   
3                    0                       0.00         4   Collision   
4                    0                       0.00        10  Electrical   

   major_regulation_change  laps_in_previous_race  race_win  new_index  
0                        0                    NaN         1         31  
1                        0                    NaN         0         58  
2                        0                    NaN         0         95  
3                        0                    NaN         0         45  
4                        0                    NaN         0         50

That looks like it has worked correctly

### Rename Columns

In [28]:
process_df = process_df.rename(columns={'Index': 'race_index'})

In order to enhance clarity within our dataset for predicting F1 race winners, I've updated the column name from 'Index' to 'race_index'

### Drop Redundant Columns

In [29]:
# Define the columns to be dropped
cols_to_drop = ["Index",
    "raceId", "resultId", "country",
    "nationality_of_circuit", "number",
    "fastestLap_ms", "status", "constructor", "positionOrder", "number.1",
    "laps", "major_regulation_change", "time", "milliseconds", "laps_completed_in_previous_races"
]

# Verify that the columns exist in the DataFrame before dropping them
existing_cols_to_drop = [col for col in cols_to_drop if col in process_df.columns]

# Drop the verified columns from the DataFrame
process_df.drop(columns=existing_cols_to_drop, axis=1, inplace=True)

I've removed unnecessary columns from our F1 race dataset, such as raceId, resultId, and others listed, to streamline and focus my predictive modeling efforts, ensuring we're working with the most relevant data for the project.

### Changing Data Types

In [30]:
process_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11235 entries, 0 to 11234
Data columns (total 22 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   race_index                  11235 non-null  int64  
 1   year                        11235 non-null  int64  
 2   race                        11235 non-null  object 
 3   driverId                    11235 non-null  int64  
 4   driver_name                 11235 non-null  object 
 5   F2_champion                 11235 non-null  int64  
 6   Former_F1_World_Champion    11235 non-null  int64  
 7   Nationality                 11235 non-null  object 
 8   home_race                   11235 non-null  int64  
 9   constructorId               11235 non-null  int64  
 10  engine_manufacturer         11235 non-null  object 
 11  constructor_nationality     11235 non-null  object 
 12  starting_grid_position      11235 non-null  int64  
 13  points                      112

The F1 dataset, consisting of 11,235 entries and 21 columns, generally has appropriate data types, with identifiers and categorical variables as integers (`int64`) and strings (`object`), and numeric scores and counters as floating-point numbers (`float64`). However, there are areas needing attention: binary columns like `F2_champion`, `Former_F1_World_Champion`, and `race_win` could be more efficiently stored as boolean (`bool`). The `fastest_lap_from_last_race` column, containing object data types, hints at a mix of strings and numbers, possibly requiring parsing and conversion to a numeric type for any time-related analysis. Moreover, the `laps_completed_in_previous_races` column only contains null values, indicating a decision point—either to populate it with relevant data or remove it if it's not useful. Overall, while the dataset seems robust for predictive modeling, it would benefit from a detailed review to ensure data integrity and optimal memory usage.

#### Turning binary columns into Booleans

In [31]:
# Convert 'F2_champion', 'Former_F1_World_Champion', 'home_race', and 'race_win' columns to boolean type
process_df['F2_champion'] = process_df['F2_champion'].astype(bool)
process_df['Former_F1_World_Champion'] = process_df['Former_F1_World_Champion'].astype(bool)
process_df['home_race'] = process_df['home_race'].astype(bool)
process_df['race_win'] = process_df['race_win'].astype(bool)

# Verify the changes
print(process_df.dtypes)


race_index                      int64
year                            int64
race                           object
driverId                        int64
driver_name                    object
F2_champion                      bool
Former_F1_World_Champion         bool
Nationality                    object
home_race                        bool
constructorId                   int64
engine_manufacturer            object
constructor_nationality        object
starting_grid_position          int64
points                        float64
points_in_previous_race       float64
fastest_lap_from_last_race     object
number_of_pit_stops             int64
average_time_lost_in_pits     float64
statusId                        int64
laps_in_previous_race         float64
race_win                         bool
new_index                       int64
dtype: object


#### Reviewing fastest_lap_from_last_race - not sure why this is an object

In [33]:
print(process_df['fastest_lap_from_last_race'].unique())


[nan '86005' '83003' '86000' '83004' '87008' '84004' '86003' 'Not Found'
 '83008' '87004' '82005' '83000' '83002' '81009' '84001' '83007' '81000'
 '84007' '88005' '89008' '88008' '101007' '96009' '94008' '94006' '95004'
 '95003' '91005' '94001' '96008' '94004' '91003' '92001' '94003' '92009'
 '95001' '94000' '93001' '98005' '97002' '120007' '96001' '90000' '96005'
 '98009' '107001' '93005' '105006' '90001' '90006' '90007' '91009' '91001'
 '106007' '92008' '94005' '93004' '100008' '86002' '88009' '86009' '87005'
 '88002' '87000' '86007' '90009' '85004' '89000' '86004' '85008' '85005'
 '86008' '87003' '92004' '90004' '87007' '88006' '88004' '90005' '87009'
 '90002' '85006' '92002' '98001' '93009' '93000' '96006' '89002' '91008'
 '101009' '92007' '91002' '92005' '91006' '81006' '82008' '80002' '83001'
 '81005' '83009' '81004' '81002' '91004' '95000' '90008' '95006' '98007'
 '93006' '90003' '93002' '96004' '104000' '119008' '116000' '114005'
 '122002' '111007' '113008' '111008' '111006' '1

Looking at this, they should be an integer as they are recorded in Millisecond. I will turn this column into an integret

In [35]:
# Replace 'Not Found' with NaN
process_df['fastest_lap_from_last_race'] = process_df['fastest_lap_from_last_race'].replace('Not Found', np.nan)

# Convert the column to Pandas nullable integer type
process_df['fastest_lap_from_last_race'] = process_df['fastest_lap_from_last_race'].astype('Int64')

### Dropping Duplicates

In [36]:
duplicate_rows = process_df[process_df.duplicated()]
duplicate_rows

Empty DataFrame
Columns: [race_index, year, race, driverId, driver_name, F2_champion, Former_F1_World_Champion, Nationality, home_race, constructorId, engine_manufacturer, constructor_nationality, starting_grid_position, points, points_in_previous_race, fastest_lap_from_last_race, number_of_pit_stops, average_time_lost_in_pits, statusId, laps_in_previous_race, race_win, new_index]
Index: []

This snippet shows a step in data preprocessing where we check for duplicate rows in our dataset. The empty DataFrame, `duplicate_rows`, shows (`0 rows x 36 columns`), indicates that there are no duplicate entries in our dataset. This is a positive sign as it means we have a unique set of data points to work with, which is crucial for building a reliable model to predict F1 race winners.

### Feature Engineering / Transformation

#### constructor_points_at_stage_of_season

In [37]:
constructor_points_sum_df = process_df.copy()

In [38]:
# First, ensure 'race_index', 'year', and 'constructorId' are sorted in the order we want to process them
constructor_points_sum_df = constructor_points_sum_df.sort_values(by=['year', 'race_index', 'constructorId'])

# Initialize a new column for corrected constructorId points
constructor_points_sum_df['corrected_constructorId_points'] = 0

# Use a temporary DataFrame to assist with the cumulative sum calculation
temp_df = constructor_points_sum_df.groupby(['year', 'race_index', 'constructorId'])['points'].sum().groupby(level=[0, 2]).cumsum().reset_index()

# Merge this temporary DataFrame back to the original sorted DataFrame
# This step ensures each driver for the constructorId at that race_index sees the summed points on the constructorId level
df_merged = pd.merge(constructor_points_sum_df, temp_df, on=['year', 'race_index', 'constructorId'], how='left')

# The merged DataFrame now has an additional column with the cumulative points which needs to be renamed and checked
df_merged = df_merged.rename(columns={'points_y': 'constructorId_points_at_stage_of_season', 'points_x': 'points'})

# Drop the previously incorrectly calculated column
df_merged.drop(columns=['corrected_constructorId_points'], inplace=True)

In [39]:
# Check the first few rows to ensure the new column has been correctly calculated
df_merged[['race_index', 'driver_name', 'year', 'constructorId', 'points', 'constructorId_points_at_stage_of_season']].head(50)

race_index             driver_name  year  constructorId  points  \
0            1           Mika Häkkinen  1995              1     3.0   
1            1           Mark Blundell  1995              1     1.0   
2            1              Damon Hill  1995              3     0.0   
3            1         David Coulthard  1995              3     6.0   
4            1          Gerhard Berger  1995              6     4.0   
5            1              Jean Alesi  1995              6     2.0   
6            1   Heinz-Harald Frentzen  1995             15     0.0   
7            1         Karl Wendlinger  1995             15     0.0   
8            1            Eddie Irvine  1995             17     0.0   
9            1      Rubens Barrichello  1995             17     0.0   
10           1       Pierluigi Martini  1995             18     0.0   
11           1             Luca Badoer  1995             18     0.0   
12           1      Michael Schumacher  1995             22    10.0   
13           1          Johnny Herbert  1995             22     0.0   
14           1           Ukyo Katayama  1995             25     0.0   
15           1               Mika Salo  1995             25     0.0   
16           1           Olivier Panis  1995             27     0.0   
17           1            Aguri Suzuki  1995             27     0.0   
18           1          Roberto Moreno  1995             28     0.0   
19           1             Pedro Diniz  1995             28     0.0   
20           1       Gianni Morbidelli  1995             29     0.0   
21           1              Taki Inoue  1995             29     0.0   
22           1         Bertrand Gachot  1995             30     0.0   
23           1       Andrea Montermini  1995             30     0.0   
24           1  Domenico Schiattarella  1995             31     0.0   
25           1          Jos Verstappen  1995             31     0.0   
26           2           Mark Blundell  1995              1     0.0   
27           2           Mika Häkkinen  1995              1     0.0   
28           2         David Coulthard  1995              3     0.0   
29           2              Damon Hill  1995              3    10.0   
30           2              Jean Alesi  1995              6     6.0   
31           2          Gerhard Berger  1995              6     1.0   
32           2         Karl Wendlinger  1995             15     0.0   
33           2   Heinz-Harald Frentzen  1995             15     2.0   
34           2      Rubens Barrichello  1995             17     0.0   
35           2            Eddie Irvine  1995             17     0.0   
36           2             Luca Badoer  1995             18     0.0   
37           2       Pierluigi Martini  1995             18     0.0   
38           2      Michael Schumacher  1995             22     4.0   
39           2          Johnny Herbert  1995             22     3.0   
40           2               Mika Salo  1995             25     0.0   
41           2           Ukyo Katayama  1995             25     0.0   
42           2            Aguri Suzuki  1995             27     0.0   
43           2           Olivier Panis  1995             27     0.0   
44           2             Pedro Diniz  1995             28     0.0   
45           2          Roberto Moreno  1995             28     0.0   
46           2              Taki Inoue  1995             29     0.0   
47           2       Gianni Morbidelli  1995             29     0.0   
48           2         Bertrand Gachot  1995             30     0.0   
49           2       Andrea Montermini  1995             30     0.0   

    constructorId_points_at_stage_of_season  
0                                       4.0  
1                                       4.0  
2                                       6.0  
3                                       6.0  
4                                       6.0  
5                                       6.0  
6                                       0.0  
7             

In [40]:
# Check the first few rows to ensure the new column has been correctly calculated
df_merged[['race_index', 'driver_name', 'year', 'constructorId', 'points', 'constructorId_points_at_stage_of_season']].tail(50)

race_index       driver_name  year  constructorId  points  \
11185         525      Lance Stroll  2023            117     0.0   
11186         525   Fernando Alonso  2023            117     6.0   
11187         525    George Russell  2023            131    10.0   
11188         525    Lewis Hamilton  2023            131    15.0   
11189         525   Kevin Magnussen  2023            210     0.0   
11190         525   Nico Hülkenberg  2023            210     0.0   
11191         525     Nyck de Vries  2023            213     0.0   
11192         525      Yuki Tsunoda  2023            213     0.0   
11193         525      Pierre Gasly  2023            214     0.0   
11194         525      Esteban Ocon  2023            214     0.0   
11195         526     Oscar Piastri  2023              1    10.0   
11196         526      Lando Norris  2023              1    18.0   
11197         526    Logan Sargeant  2023              3     0.0   
11198         526   Alexander Albon  2023              3     0.0   
11199         526      Carlos Sainz  2023              6     4.0   
11200         526   Charles Leclerc  2023              6     6.0   
11201         526    Max Verstappen  2023              9    26.0   
11202         526      Sergio Pérez  2023              9    15.0   
11203         526   Valtteri Bottas  2023             51     0.0   
11204         526       Guanyu Zhou  2023             51     0.0   
11205         526      Lance Stroll  2023            117     1.0   
11206         526   Fernando Alonso  2023            117     2.0   
11207         526    George Russell  2023            131     8.0   
11208         526    Lewis Hamilton  2023            131    12.0   
11209         526   Kevin Magnussen  2023            210     0.0   
11210         526   Nico Hülkenberg  2023            210     0.0   
11211         526  Daniel Ricciardo  2023            213     0.0   
11212         526      Yuki Tsunoda  2023            213     0.0   
11213         526      Pierre Gasly  2023            214     0.0   
11214         526      Esteban Ocon  2023            214     0.0   
11215         527      Lando Norris  2023              1     6.0   
11216         527     Oscar Piastri  2023              1     0.0   
11217         527    Logan Sargeant  2023              3     0.0   
11218         527   Alexander Albon  2023              3     0.0   
11219         527   Charles Leclerc  2023              6    15.0   
11220         527      Carlos Sainz  2023              6     0.0   
11221         527    Max Verstappen  2023              9    25.0   
11222         527      Sergio Pérez  2023              9    18.0   
11223         527   Valtteri Bottas  2023             51     0.0   
11224         527       Guanyu Zhou  2023             51     0.0   
11225         527      Lance Stroll  2023            117     2.0   
11226         527   Fernando Alonso  2023            117    10.0   
11227         527    Lewis Hamilton  2023            131    13.0   
11228         527    George Russell  2023            131     8.0   
11229         527   Kevin Magnussen  2023            210     0.0   
11230         527   Nico Hülkenberg  2023            210     0.0   
11231         527  Daniel Ricciardo  2023            213     0.0   
11232         527      Yuki Tsunoda  2023            213     1.0   
11233         527      Pierre Gasly  2023            214     0.0   
11234         527      Esteban Ocon  2023            214     4.0   

       constructorId_points_at_stage_of_season  
11185                                    168.0  
11186                                    168.0  
11187                                    195.0  
11188                                    195.0  
11189                                      8.0  
11190                                      8.0  
11191                                      2.0  
11192                                      2.0  
11193                                     45.0  
11194                                     45.0  

That seems to have worked - now we have a column that has the cumulative points for the constructor at this stage of the season for each driver.

#### driver_points_at_stage_of_season

In [41]:
driver_points_sum_df = df_merged.copy()

In [42]:
# Use a temporary DataFrame to assist with the cumulative sum calculation for drivers
temp_driver_df = driver_points_sum_df.groupby(['year', 'race_index', 'driver_name'])['points'].sum().groupby(level=[0, 2]).cumsum().reset_index()

# Merge this temporary DataFrame back to the original DataFrame
# This step ensures each driver sees the summed points at that stage of the season
df_merged_with_driver_points = pd.merge(driver_points_sum_df, temp_driver_df, on=['year', 'race_index', 'driver_name'], how='left')

# The merged DataFrame now has an additional column with the cumulative points which needs to be renamed and checked
df_merged_with_driver_points = df_merged_with_driver_points.rename(columns={'points_y': 'driver_points_at_stage_of_season', 'points_x': 'points'})

In [43]:
# Check the first few rows to ensure the new column has been correctly calculated
df_merged_with_driver_points[['race_index', 'year', 'driver_name', 'points', 'driver_points_at_stage_of_season']].head(50)

race_index  year             driver_name  points  \
0            1  1995           Mika Häkkinen     3.0   
1            1  1995           Mark Blundell     1.0   
2            1  1995              Damon Hill     0.0   
3            1  1995         David Coulthard     6.0   
4            1  1995          Gerhard Berger     4.0   
5            1  1995              Jean Alesi     2.0   
6            1  1995   Heinz-Harald Frentzen     0.0   
7            1  1995         Karl Wendlinger     0.0   
8            1  1995            Eddie Irvine     0.0   
9            1  1995      Rubens Barrichello     0.0   
10           1  1995       Pierluigi Martini     0.0   
11           1  1995             Luca Badoer     0.0   
12           1  1995      Michael Schumacher    10.0   
13           1  1995          Johnny Herbert     0.0   
14           1  1995           Ukyo Katayama     0.0   
15           1  1995               Mika Salo     0.0   
16           1  1995           Olivier Panis     0.0   
17           1  1995            Aguri Suzuki     0.0   
18           1  1995          Roberto Moreno     0.0   
19           1  1995             Pedro Diniz     0.0   
20           1  1995       Gianni Morbidelli     0.0   
21           1  1995              Taki Inoue     0.0   
22           1  1995         Bertrand Gachot     0.0   
23           1  1995       Andrea Montermini     0.0   
24           1  1995  Domenico Schiattarella     0.0   
25           1  1995          Jos Verstappen     0.0   
26           2  1995           Mark Blundell     0.0   
27           2  1995           Mika Häkkinen     0.0   
28           2  1995         David Coulthard     0.0   
29           2  1995              Damon Hill    10.0   
30           2  1995              Jean Alesi     6.0   
31           2  1995          Gerhard Berger     1.0   
32           2  1995         Karl Wendlinger     0.0   
33           2  1995   Heinz-Harald Frentzen     2.0   
34           2  1995      Rubens Barrichello     0.0   
35           2  1995            Eddie Irvine     0.0   
36           2  1995             Luca Badoer     0.0   
37           2  1995       Pierluigi Martini     0.0   
38           2  1995      Michael Schumacher     4.0   
39           2  1995          Johnny Herbert     3.0   
40           2  1995               Mika Salo     0.0   
41           2  1995           Ukyo Katayama     0.0   
42           2  1995            Aguri Suzuki     0.0   
43           2  1995           Olivier Panis     0.0   
44           2  1995             Pedro Diniz     0.0   
45           2  1995          Roberto Moreno     0.0   
46           2  1995              Taki Inoue     0.0   
47           2  1995       Gianni Morbidelli     0.0   
48           2  1995         Bertrand Gachot     0.0   
49           2  1995       Andrea Montermini     0.0   

    driver_points_at_stage_of_season  
0                                3.0  
1                                1.0  
2                                0.0  
3                                6.0  
4                                4.0  
5                                2.0  
6                                0.0  
7                                0.0  
8                                0.0  
9                                0.0  
10                               0.0  
11                               0.0  
12                              10.0  
13                               0.0  
14                               0.0  
15                               0.0  
16                               0.0  
17                               0.0  
18                               0.0  
19                               0.0  
20                               0.0  
21                               0.0  
22                               0.0  
23                               0.0  
24                               0.0  
25                               0.0  
26                               1.0  
27                               3.0  
28              

In [44]:
# Check the first few rows to ensure the new column has been correctly calculated
df_merged_with_driver_points[['race_index', 'year', 'driver_name', 'points', 'driver_points_at_stage_of_season']].tail(50)

race_index  year       driver_name  points  \
11185         525  2023      Lance Stroll     0.0   
11186         525  2023   Fernando Alonso     6.0   
11187         525  2023    George Russell    10.0   
11188         525  2023    Lewis Hamilton    15.0   
11189         525  2023   Kevin Magnussen     0.0   
11190         525  2023   Nico Hülkenberg     0.0   
11191         525  2023     Nyck de Vries     0.0   
11192         525  2023      Yuki Tsunoda     0.0   
11193         525  2023      Pierre Gasly     0.0   
11194         525  2023      Esteban Ocon     0.0   
11195         526  2023     Oscar Piastri    10.0   
11196         526  2023      Lando Norris    18.0   
11197         526  2023    Logan Sargeant     0.0   
11198         526  2023   Alexander Albon     0.0   
11199         526  2023      Carlos Sainz     4.0   
11200         526  2023   Charles Leclerc     6.0   
11201         526  2023    Max Verstappen    26.0   
11202         526  2023      Sergio Pérez    15.0   
11203         526  2023   Valtteri Bottas     0.0   
11204         526  2023       Guanyu Zhou     0.0   
11205         526  2023      Lance Stroll     1.0   
11206         526  2023   Fernando Alonso     2.0   
11207         526  2023    George Russell     8.0   
11208         526  2023    Lewis Hamilton    12.0   
11209         526  2023   Kevin Magnussen     0.0   
11210         526  2023   Nico Hülkenberg     0.0   
11211         526  2023  Daniel Ricciardo     0.0   
11212         526  2023      Yuki Tsunoda     0.0   
11213         526  2023      Pierre Gasly     0.0   
11214         526  2023      Esteban Ocon     0.0   
11215         527  2023      Lando Norris     6.0   
11216         527  2023     Oscar Piastri     0.0   
11217         527  2023    Logan Sargeant     0.0   
11218         527  2023   Alexander Albon     0.0   
11219         527  2023   Charles Leclerc    15.0   
11220         527  2023      Carlos Sainz     0.0   
11221         527  2023    Max Verstappen    25.0   
11222         527  2023      Sergio Pérez    18.0   
11223         527  2023   Valtteri Bottas     0.0   
11224         527  2023       Guanyu Zhou     0.0   
11225         527  2023      Lance Stroll     2.0   
11226         527  2023   Fernando Alonso    10.0   
11227         527  2023    Lewis Hamilton    13.0   
11228         527  2023    George Russell     8.0   
11229         527  2023   Kevin Magnussen     0.0   
11230         527  2023   Nico Hülkenberg     0.0   
11231         527  2023  Daniel Ricciardo     0.0   
11232         527  2023      Yuki Tsunoda     1.0   
11233         527  2023      Pierre Gasly     0.0   
11234         527  2023      Esteban Ocon     4.0   

       driver_points_at_stage_of_season  
11185                              38.0  
11186                             130.0  
11187                              76.0  
11188                             119.0  
11189                               2.0  
11190                               6.0  
11191                               0.0  
11192                               2.0  
11193                              16.0  
11194                              29.0  
11195                              27.0  
11196                              60.0  
11197                               0.0  
11198                              11.0  
11199                              77.0  
11200                              73.0  
11201                             267.0  
11202                             156.0  
11203                               5.0  
11204                               4.0  
11205                              39.0  
11206                             132.0  
11207                              84.0  
11208                             131.0  
11209                               2.0  
11210                               6.0  
11211                               0.0  
11212                               2.0  
11213                              16.0  
11214                              29.0  
1

That seems to have worked - now we have a column that has the cumulative points for the driver at this stage of the season for each driver. Now we will review the where we are with the dataframe again.

In [48]:
df_merged_with_driver_points.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11235 entries, 0 to 11234
Data columns (total 19 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   race_index                               11235 non-null  int64  
 1   year                                     11235 non-null  int64  
 2   race                                     11235 non-null  object 
 3   driverId                                 11235 non-null  int64  
 4   F2_champion                              11235 non-null  bool   
 5   Former_F1_World_Champion                 11235 non-null  bool   
 6   Nationality                              11235 non-null  object 
 7   home_race                                11235 non-null  bool   
 8   constructorId                            11235 non-null  int64  
 9   engine_manufacturer                      11235 non-null  object 
 10  constructor_nationality                  11235

In [49]:
df_merged_with_driver_points.head()

race_index  year                  race  driverId  F2_champion  \
0           1  1995  Brazilian Grand Prix        57        False   
1           1  1995  Brazilian Grand Prix        87        False   
2           1  1995  Brazilian Grand Prix        71        False   
3           1  1995  Brazilian Grand Prix        14        False   
4           1  1995  Brazilian Grand Prix        77        False   

   Former_F1_World_Champion Nationality  home_race  constructorId  \
0                     False     Finnish      False              1   
1                     False     British      False              1   
2                      True     British      False              3   
3                     False     British      False              3   
4                     False    Austrian      False              6   

  engine_manufacturer constructor_nationality  starting_grid_position  \
0            Mercedes                 British                       7   
1            Mercedes                 British                       9   
2             Renault                 British                       1   
3             Renault                 British                       3   
4             Ferrari                 Italian                       5   

   points_in_previous_race  fastest_lap_from_last_race  laps_in_previous_race  \
0                      NaN                        <NA>                    NaN   
1                      NaN                        <NA>                    NaN   
2                      NaN                        <NA>                    NaN   
3                      NaN                        <NA>                    NaN   
4                      NaN                        <NA>                    NaN   

   race_win  new_index  constructorId_points_at_stage_of_season  \
0     False         58                                      4.0   
1     False         88                                      4.0   
2     False         72                                      6.0   
3     False         15                                      6.0   
4     False         78                                      6.0   

   driver_points_at_stage_of_season  
0                               3.0  
1                               1.0  
2                               0.0  
3                               6.0  
4                               4.0

In [51]:
# Define the columns to be dropped
cols_to_drop = [
    "driver_name", "resultId", "number_of_pit_stops", "average_time_lost_in_pits",
    "statusId", "laps_in_previous_races", "points", "fastest_lap_from_last_race"
]

NB - laps_completed_in_previous_races, I have dropped for the time being. It was full of null values. I will run the first models and use thenm as benchmarks and then see if I can improve the scores. 
The other columns I didn't think were helpful for the following reasons
- `number_of_pit_stops` & `average_time_lost_in_pits` - as we are trying to predict the race winner before the race, this information isn't very helpful at this stage. It may be interesting to see how last year's pit strategy at a particular race will impact next year's race, but it's not so helpful in it's current form.
- `points` - this isn't helpful, as we want to predict forwards and this is captured from `constructorId_points_at_stage_of_season` & `driver_points_at_stage_of_season` columns
- `fastest_lap_from_last_race` and `laps_in_previous_races` are currently missing 

In [52]:

# Verify that the columns exist in the DataFrame before dropping them
existing_cols_to_drop = [col for col in cols_to_drop if col in df_merged_with_driver_points.columns]

# Drop the verified columns from the DataFrame
df_merged_with_driver_points.drop(columns=existing_cols_to_drop, axis=1, inplace=True)

In [53]:
df_merged_with_driver_points.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11235 entries, 0 to 11234
Data columns (total 18 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   race_index                               11235 non-null  int64  
 1   year                                     11235 non-null  int64  
 2   race                                     11235 non-null  object 
 3   driverId                                 11235 non-null  int64  
 4   F2_champion                              11235 non-null  bool   
 5   Former_F1_World_Champion                 11235 non-null  bool   
 6   Nationality                              11235 non-null  object 
 7   home_race                                11235 non-null  bool   
 8   constructorId                            11235 non-null  int64  
 9   engine_manufacturer                      11235 non-null  object 
 10  constructor_nationality                  11235

### One Hot Encoding

In [54]:
# Perform one-hot encoding on the specified columns
df_with_dummies = pd.get_dummies(df_merged_with_driver_points, columns=['race','engine_manufacturer', 'constructor_nationality', 'Nationality'])

# Now df_with_dummies contains the original data along with the one-hot encoded columns


In [58]:
df_with_dummies.head()

race_index  year  driverId  F2_champion  Former_F1_World_Champion  \
0           1  1995        57        False                     False   
1           1  1995        87        False                     False   
2           1  1995        71        False                      True   
3           1  1995        14        False                     False   
4           1  1995        77        False                     False   

   home_race  constructorId  starting_grid_position  points_in_previous_race  \
0      False              1                       7                      NaN   
1      False              1                       9                      NaN   
2      False              3                       1                      NaN   
3      False              3                       3                      NaN   
4      False              6                       5                      NaN   

   laps_in_previous_race  race_win  new_index  \
0                    NaN     False         58   
1                    NaN     False         88   
2                    NaN     False         72   
3                    NaN     False         15   
4                    NaN     False         78   

   constructorId_points_at_stage_of_season  driver_points_at_stage_of_season  \
0                                      4.0                               3.0   
1                                      4.0                               1.0   
2                                      6.0                               0.0   
3                                      6.0                               6.0   
4                                      6.0                               4.0   

   race_70th Anniversary Grand Prix  race_Abu Dhabi Grand Prix  \
0                             False                      False   
1                             False                      False   
2                             False                      False   
3                             False                      False   
4                             False                      False   

   race_Argentine Grand Prix  race_Australian Grand Prix  \
0                      False                       False   
1                      False                       False   
2                      False                       False   
3                      False                       False   
4                      False                       False   

   race_Austrian Grand Prix  race_Azerbaijan Grand Prix  \
0                     False                       False   
1                     False                       False   
2                     False                       False   
3                     False                       False   
4                     False                       False   

   race_Bahrain Grand Prix  race_Belgian Grand Prix  \
0                    False                    False   
1                    False                    False   
2                    False                    False   
3                    False                    False   
4                    False                    False   

   race_Brazilian Grand Prix  race_British Grand Prix  \
0                       True                    False   
1                       True                    False   
2                       True                    False   
3                       True                    False   
4                       True                    False   

   race_Canadian Grand Prix  race_Chinese Grand Prix  race_Dutch Grand Prix  \
0                     False                    False                  False   
1                     False                    False                  False   
2                     False                    False                  False   
3                     False                    False                  False   
4                     False                    False                  False   

   race_Eifel Grand Prix  race_Emilia Romagna Grand

In our pursuit to predict F1 race winners, it's critical to utilize data that is both relevant and complete. For the inaugural race, indexed as 1 in our dataset, there is no preceding race data available! This absence of prior race information creates gaps in our data (reflected as NaN values), which can mislead our predictive models and skew results. To ensure the integrity and efficacy of our analysis, we remove this first race from our dataset, along with any rows marred by NaNs, thus aligning our data foundation with the solid ground of complete historical context and statistical validity.

In [59]:
# Remove all rows with race_index equal to 1
df_with_dummies = df_with_dummies[df_with_dummies['race_index'] != 1]

In [60]:
df_with_dummies.head()

race_index  year  driverId  F2_champion  Former_F1_World_Champion  \
26           2  1995        87        False                     False   
27           2  1995        57        False                     False   
28           2  1995        14        False                     False   
29           2  1995        71        False                      True   
30           2  1995        55        False                     False   

    home_race  constructorId  starting_grid_position  points_in_previous_race  \
26      False              1                      17                      1.0   
27      False              1                       5                      3.0   
28      False              3                       1                      6.0   
29      False              3                       2                      0.0   
30      False              6                       6                      2.0   

    laps_in_previous_race  race_win  new_index  \
26                   70.0     False         89   
27                   70.0     False         59   
28                   71.0     False         16   
29                   30.0      True         73   
30                   70.0     False         57   

    constructorId_points_at_stage_of_season  driver_points_at_stage_of_season  \
26                                      4.0                               1.0   
27                                      4.0                               3.0   
28                                     16.0                               6.0   
29                                     16.0                              10.0   
30                                     13.0                               8.0   

    race_70th Anniversary Grand Prix  race_Abu Dhabi Grand Prix  \
26                             False                      False   
27                             False                      False   
28                             False                      False   
29                             False                      False   
30                             False                      False   

    race_Argentine Grand Prix  race_Australian Grand Prix  \
26                       True                       False   
27                       True                       False   
28                       True                       False   
29                       True                       False   
30                       True                       False   

    race_Austrian Grand Prix  race_Azerbaijan Grand Prix  \
26                     False                       False   
27                     False                       False   
28                     False                       False   
29                     False                       False   
30                     False                       False   

    race_Bahrain Grand Prix  race_Belgian Grand Prix  \
26                    False                    False   
27                    False                    False   
28                    False                    False   
29                    False                    False   
30                    False                    False   

    race_Brazilian Grand Prix  race_British Grand Prix  \
26                      False                    False   
27                      False                    False   
28                      False                    False   
29                      False                    False   
30                      False                    False   

    race_Canadian Grand Prix  race_Chinese Grand Prix  race_Dutch Grand Prix  \
26                     False                    False                  False   
27                     False                    False                  False   
28                     False                    False                  False   
29                     False                    False                  False   
30                     False                    False                  Fa

Now I will check to see if there are any more NaN values

In [62]:
# Calculate the total number of NaN values in the DataFrame
total_nan_values = df_with_dummies.isna().sum().sum()

# Print the total number of NaN values
print(f"The total number of NaN values in the DataFrame is: {total_nan_values}")


The total number of NaN values in the DataFrame is: 576


In [64]:
# Filter the DataFrame to only include rows with at least one NaN value
rows_with_nan = df_with_dummies[df_with_dummies.isna().any(axis=1)]

# Display those rows
rows_with_nan


race_index  year  driverId  F2_champion  Former_F1_World_Champion  \
53              3  1995        95        False                     False   
95              4  1995        84        False                     False   
105             5  1995        87        False                     False   
110             5  1995        96        False                     False   
198             8  1995        97        False                     False   
...           ...   ...       ...          ...                       ...   
10995         516  2023       857         True                     False   
10998         516  2023       858        False                     False   
11010         516  2023       807         True                     False   
11011         516  2023       856         True                     False   
11211         526  2023       817        False                     False   

       home_race  constructorId  starting_grid_position  \
53         False              1                       9   
95         False             27                      11   
105        False              1                      10   
110        False             15                      19   
198        False             29                      17   
...          ...            ...                     ...   
10995      False              1                      18   
10998      False              3                      16   
11010      False            210                      10   
11011      False            213                      19   
11211      False            213                      13   

       points_in_previous_race  laps_in_previous_race  race_win  new_index  \
53                         NaN                    NaN     False         98   
95                         NaN                    NaN     False         88   
105                        NaN                    NaN     False         92   
110                        NaN                    NaN     False        101   
198                        NaN                    NaN     False        105   
...                        ...                    ...       ...        ...   
10995                      NaN                    NaN     False       1373   
10998                      NaN                    NaN     False       1374   
11010                      NaN                    NaN     False       1323   
11011                      NaN                    NaN     False       1372   
11211                      NaN                    NaN     False       1343   

       constructorId_points_at_stage_of_season  \
53                                         6.0   
95                                         1.0   
105                                        8.0   
110                                        4.0   
198                                        1.0   
...                                        ...   
10995                                      0.0   
10998                                      1.0   
11010                                      0.0   
11011                                      0.0   
11211                                      2.0   

       driver_points_at_stage_of_season  race_70th Anniversary Grand Prix  \
53                                  0.0                             False   
95                                  0.0                             False   
105                                 3.0                             False   
110                                 0.0                             False   
198                                 0.0                             False   
...                                 ...                               ...   
10995                               0.0                             False   
10998                               0.0                             False   
11010                               0.0                             False   
11011                               0.0                             False   


In [66]:
# Calculate the total number of rows in the DataFrame
total_rows = df_with_dummies.shape[0]

# Calculate the number of rows that contain at least one NaN value
rows_with_nan = df_with_dummies.isna().any(axis=1).sum()

# Calculate the percentage of rows with at least one NaN value
percentage_with_nan = (rows_with_nan / total_rows) * 100

# Print the percentage
print(f"Percentage of rows with at least one NaN value: {percentage_with_nan:.2f}%")


Percentage of rows with at least one NaN value: 2.57%


At this is less than 5%, I will drop these rows.

In [67]:
# Drop all rows that contain any NaN values
df_cleaned = df_with_dummies.dropna()

# Verify changes by displaying the new DataFrame
print(df_cleaned)


       race_index  year  driverId  F2_champion  Former_F1_World_Champion  \
26              2  1995        87        False                     False   
27              2  1995        57        False                     False   
28              2  1995        14        False                     False   
29              2  1995        71        False                      True   
30              2  1995        55        False                     False   
...           ...   ...       ...          ...                       ...   
11230         527  2023       807         True                     False   
11231         527  2023       817        False                     False   
11232         527  2023       852        False                     False   
11233         527  2023       842         True                     False   
11234         527  2023       839        False                     False   

       home_race  constructorId  starting_grid_position  \
26         False            

This looks to have worked

In [68]:
df_cleaned.to_csv('C:/Users/Alex/OneDrive/BrainStation/Data_Science_Bootcamp/Capstone_Project/capstone-Aboard89/model_data.csv', index=False)

## **Conclusion**


In conclusion, the data preparation for the F1 Win Prediction Project involved several crucial preprocessing steps to ensure the data was primed for analysis and modeling:

1. **Creating a New Index**: A new index column was created by adding the 'Index' column and the 'driverId' column to track predictions.

2. **Renaming Columns**: The 'Index' column was renamed to 'race_index'.

3. **Dropping Redundant Columns**: Specific columns that were deemed unnecessary for analysis were dropped from the DataFrame. 

4. **Verifying Data Types**: The types of certain columns were checked to ensure they were correct, although the specific commands for type conversion were commented out and thus not executed.

5. **Dropping Duplicates**: Code to drop duplicate rows was included but commented out, hence not executed.

6. **Handling Data Ranges**: Code to handle unreasonable data ranges was present but commented out.

7. **Feature Engineering - Cumulative Constructor Points**: The DataFrame was sorted, and a new column for constructor points at the stage of the season was calculated using groupby and merge operations.

8. **Feature Engineering - Cumulative Driver Points**: Similarly, a new column for driver points at the stage of the season was calculated using temporary DataFrames to store cumulative sums, which were then merged back into the original DataFrame.

9. **Dropping Verified Columns**: Additional redundant columns were dropped after verifying their existence in the DataFrame.

10. **Exporting Processed Data**: The processed DataFrame was then exported to a CSV file.

11. **One Hot Encoding**: One-hot encoding was performed on specific categorical columns using pd.get_dummies.

12. **Reviewing Column Data Types**: The unique values of the 'fastest_lap_from_last_race' column were printed out, suggesting a review of why it was of the 'object' data type.